src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,529 @@
+                writer.Write("localtime", options.CurrentLocalTime.ToString(CultureInfo.InvariantCulture));
+
+                // When using /deterministic- the compiler will *always* emit different binaries hence the 
+                // key we generate here also must be different. We cannot depend on the `localtime` property
+                // to provide this as the same compilation can occur on different machines at the same 
+                // time. Force the issue here.
+                writer.Write("nondeterministicMvid", GetGuidValue(Guid.NewGuid()));
+            }
+
+            // Values which do not impact build success / failure
+            // - ConcurrentBuild
+            // - MetadataImportOptions:
+            // - Options.Features: deprecated
+            // 
+
+            // Not really options but they can impact compilation so we record the types. For the majority
+            // of compilations this is roughly the equivalent of recording the compiler version but it 
+            // could differ when customers host the compiler via the API.
+            writer.WriteKey("extensions");
+            writer.WriteObjectStart();
+
+            WriteType(writer, "syntaxTreeOptionsProvider", options.SyntaxTreeOptionsProvider?.GetType());
+            WriteType(writer, "metadataReferenceResolver", options.MetadataReferenceResolver?.GetType());
+            WriteType(writer, "xmlReferenceResolver", options.XmlReferenceResolver?.GetType());
+            WriteType(writer, "sourceReferenceResolver", options.SourceReferenceResolver?.GetType());
+            WriteType(writer, "strongNameProvider", options.StrongNameProvider?.GetType());
+            WriteType(writer, "assemblyIdentityComparer", options.AssemblyIdentityComparer?.GetType());
+            writer.WriteObjectEnd();
+        }
+
+        protected void WriteParseOptions(JsonWriter writer, ParseOptions parseOptions)
+        {
+            writer.WriteObjectStart();
+            WriteParseOptionsCore(writer, parseOptions);
+            writer.WriteObjectEnd();
+        }
+
+        protected virtual void WriteParseOptionsCore(JsonWriter writer, ParseOptions parseOptions)
+        {
+            writer.Write("kind", parseOptions.Kind);
+            writer.Write("specifiedKind", parseOptions.SpecifiedKind);
+            writer.Write("documentationMode", parseOptions.DocumentationMode);
+            writer.Write("language", parseOptions.Language);
+
+            writer.WriteKey("features");
+            var features = parseOptions.Features;
+            if (features.Count > 0)
+            {
+                writer.WriteObjectStart();
+                foreach (var key in features.Keys.OrderBy(StringComparer.Ordinal))
+                {
+                    writer.Write(key, features[key]);
+                }
+                writer.WriteObjectEnd();
+            }
+            else
+            {
+                writer.WriteNull();
```

Comment:
consider not special casing and just writing out an empty objeect.

Reply:
Good catch. Rest of the code works this way, missed this one.